crates/uv-pep508/src/marker/algebra.rs

```diff
@@ -421,6 +489,245 @@ impl InternerGuard<'_> {
+    /// both `sys_platform == 'win32'` and `platform_system == 'Darwin'` will never true at the
+    /// same time.
+    ///
+    /// This method thus encodes assumptions about the environment that are not guaranteed by the
+    /// PEP 508 specification alone.
+    fn exclusions(&mut self) -> NodeId {
+        /// Perform a disjunction operation between two nodes.
+        ///
+        /// This is equivalent to [`InternerGuard::or`], with the exception that it does not
+        /// incorporate knowledge from outside the marker algebra.
+        fn disjunction(guard: &mut InternerGuard<'_>, xi: NodeId, yi: NodeId) -> NodeId {
+            // We take advantage of cheap negation here and implement OR in terms
+            // of it's De Morgan complement.
+            conjunction(guard, xi.not(), yi.not()).not()
+        }
+
+        /// Perform a conjunction operation between two nodes.
+        ///
+        /// This is equivalent to [`InternerGuard::and`], with the exception that it does not
+        /// incorporate knowledge from outside the marker algebra.
+        fn conjunction(guard: &mut InternerGuard<'_>, xi: NodeId, yi: NodeId) -> NodeId {
+            if xi.is_true() {
+                return yi;
+            }
+            if yi.is_true() {
+                return xi;
+            }
+            if xi == yi {
+                return xi;
+            }
+            if xi.is_false() || yi.is_false() {
+                return NodeId::FALSE;
+            }
+            // `X and not X` is `false` by definition.
+            if xi.not() == yi {
+                return NodeId::FALSE;
+            }
+
+            let (x, y) = (guard.shared.node(xi), guard.shared.node(yi));
+
+            // Perform Shannon Expansion of the higher order variable.
+            let (func, children) = match x.var.cmp(&y.var) {
+                // X is higher order than Y, apply Y to every child of X.
+                Ordering::Less => {
+                    let children = x.children.map(xi, |node| conjunction(guard, node, yi));
+                    (x.var.clone(), children)
+                }
+                // Y is higher order than X, apply X to every child of Y.
+                Ordering::Greater => {
+                    let children = y.children.map(yi, |node| conjunction(guard, node, xi));
+                    (y.var.clone(), children)
+                }
+                // X and Y represent the same variable, merge their children.
+                Ordering::Equal => {
+                    let children = x
+                        .children
+                        .apply(xi, &y.children, yi, |x, y| conjunction(guard, x, y));
+                    (x.var.clone(), children)
+                }
+            };
+
+            // Create the output node.
+            guard.create_node(func, children)
+        }
+
+        if let Some(exclusions) = self.state.exclusions {
+            return exclusions;
+        }
```

Comment:
I assume it's fine to just use an `Option` here rather than a `OnceCell`?

Reply:
Yeah. Rust wouldn't let you mess this up.